impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -38,225 +82,312 @@ Join StrokePathGeometry::GetStrokeJoin() const {
+  CreateBevelAndGetDirection(vtx_builder, position, start_offset, end_offset);
+}
+
 // static
 StrokePathGeometry::JoinProc StrokePathGeometry::GetJoinProc(Join stroke_join) {
-  using VS = SolidFillVertexShader;
-  StrokePathGeometry::JoinProc join_proc;
   switch (stroke_join) {
     case Join::kBevel:
-      join_proc = [](VertexBufferBuilder<VS::PerVertexData>& vtx_builder,
-                     const Point& position, const Point& start_offset,
-                     const Point& end_offset, Scalar miter_limit,
-                     Scalar scale) {
-        CreateBevelAndGetDirection(vtx_builder, position, start_offset,
-                                   end_offset);
-      };
-      break;
+      return &StrokePathGeometry::CreateBevelJoin;
     case Join::kMiter:
-      join_proc = [](VertexBufferBuilder<VS::PerVertexData>& vtx_builder,
-                     const Point& position, const Point& start_offset,
-                     const Point& end_offset, Scalar miter_limit,
-                     Scalar scale) {
-        Point start_normal = start_offset.Normalize();
-        Point end_normal = end_offset.Normalize();
-
-        // 1 for no joint (straight line), 0 for max joint (180 degrees).
-        Scalar alignment = (start_normal.Dot(end_normal) + 1) / 2;
-        if (ScalarNearlyEqual(alignment, 1)) {
-          return;
-        }
+      return &StrokePathGeometry::CreateMiterJoin;
+    case Join::kRound:
+      return &StrokePathGeometry::CreateRoundJoin;
+  }
+}
 
-        Scalar dir = CreateBevelAndGetDirection(vtx_builder, position,
-                                                start_offset, end_offset);
+// static
+void StrokePathGeometry::CreateButtCap(VertexWriter& vtx_builder,
+                                       const Point& position,
+                                       const Point& offset,
+                                       Scalar scale,
+                                       bool reverse) {
+  Point orientation = offset * (reverse ? -1 : 1);
+  VS::PerVertexData vtx;
+  vtx.position = position + orientation;
+  vtx_builder.AppendVertex(vtx.position);
+  vtx.position = position - orientation;
+  vtx_builder.AppendVertex(vtx.position);
+}
 
-        Point miter_point = (start_offset + end_offset) / 2 / alignment;
-        if (miter_point.GetDistanceSquared({0, 0}) >
-            miter_limit * miter_limit) {
-          return;  // Convert to bevel when we exceed the miter limit.
-        }
+// static
+void StrokePathGeometry::CreateRoundCap(VertexWriter& vtx_builder,
+                                        const Point& position,
+                                        const Point& offset,
+                                        Scalar scale,
+                                        bool reverse) {
+  Point orientation = offset * (reverse ? -1 : 1);
+  Point forward(offset.y, -offset.x);
+  Point forward_normal = forward.Normalize();
+
+  CubicPathComponent arc;
```

Comment:
It looks like the EllipticalVertexGenerator would do what I want. Does that generate two full circles for the each center? I think to incorporate it with the existing triangle strip I'd need to do one at a time.

Reply:
That class is a very thin veil over the generator proc. The generator procs get a list of trig values (sin/cos) spaced reasonably for a circle of a given radius. The trig arrays are basically a single quarter circle quadrant of a full circle and they are iterated in various ways 4 times over to produce zig-zag triangle strips. They are the raw trig values so you still need to multiply by the distance (and potentially the direction) - the thing that varies based on the pixel size is how many of them to produce.

The `GenerateRoundCapLine` proc is close to what we need since it generates the quadrants at an angle based on a direction vector and breaks it down into 2 vectors - one along the line and one across the line both sized by the stroke width. But it generates a triangle strip instead of a poly-line so it's mostly a matter of the order in which the trigs are iterated and multiplied by the along and across vectors. The (triangle) vertices for a single line segment start at the very outward end of the round cap (the top of the semicircle) and zig zag down it until they meet the base of the cap (which is where a butt cap would just cut across) - it then does the reverse on the far round cap since it is responsible for both ends of a drawLine and the jump from one cap to the other naturally covers the body of the line segment itself.

For path tracing you only deal with one round cap at a time and since you are generating polylines instead of triangle strips you would need to iterate in a different order - basically start at the base of the semi-circle and run up one side and then down the other - iterating the trigs twice and moving along the semi-circle instead of zig-zagging down it from the top.

A couple of tricks - the symmetry of the sin/cos values means that whether you are going from 0 to 90 degrees or 90 to 0 degrees you can iterator forward over the array and just swap the sin/cos values. So, flipping from going up one side of the semicircle and then down the other side would just involve iterating forward over the trigs twice, once using cos/sin and the second time using sin/cos and changing the sign of whether the across vector is added or subtracted from the segment endpoint.

Since this will be the first (potentially only) generator proc that produces a polyline instead of a triangle strip we should probably make that clear in the name(s).

Reply:
While we do generate a polyline for the stroke, that doesn't include the caps/joins, those are appended as triangle strips as we process the polyline.

Reply:
Ah, then it gets even easier. It's more a matter of only generating half of the circle than switching the vertex ordering - for the caps. The joins would still require computing the correct subset of the trigs to use which is something none of the current generators try to do.